### PR TITLE
[WIP] Integrate in-memory object store into arrow

### PR DESCRIPTION
**Pull request by pcmoritz**
This is work in progress to integrate an in-memory object store into apache arrow.

A number of code cleanups need to happen, among others:

1. replacing uthash and utarray with STL containers by rebasing https://github.com/ray-project/ray/pull/324
2. applying the arrow linting
3. use arrow status codes throughout the code base
4. do some more modernization of the codebase (more idiomatic c++)

Comments and help is appreciated!

### CODE REVIEW

**Comment by wesm** [comment]
I will take a look through as soon as I can. With ray-project/ray#324 open, it will become tricky to "mirror in" changes from Ray, ideally we could get this to a state within a few weeks where you can use it from your thirdparty toolchain. What do you think?

**Comment by robertnishihara** [comment]
Yeah, it'd be good to get as quickly as possible to a state where we're using the object store entirely from Arrow (and testing the object store primarily with Arrow's CI).

We'll have to be careful about changes to the object store that are happening before the transition. There's ray-project/ray#324, but there are also some performance improvements that @atumanov has been working on (but there is no PR for that yet).

Changes that are made to the plasma store in Ray before this PR is merged could just be added to this PR.

**Comment by pcmoritz** [comment]
Yes I agree with @robertnishihara and @wesm. The timeline also looks good to me.

We are working on getting all the necessary changes into Ray and I'll keep this PR up to date. So it shouldn't block the work on getting this code more in line with the arrow style.

File: cpp/src/plasma/cmake/Common.cmake [hunk 1]
```diff
@@ -0,0 +1,93 @@
+# Code for compiling flatbuffers
+
+include(ExternalProject)
+include(CMakeParseArguments)
+
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11")
+
+set(FLATBUFFERS_VERSION "1.6.0")
+
+set(FLATBUFFERS_PREFIX "${CMAKE_BINARY_DIR}/flatbuffers_ep-prefix/src/flatbuffers_ep-install")
```
Comment (robertnishihara) [inline cpp/src/plasma/cmake/Common.cmake]: FYI, we're pulling in our own copy of flatbuffers here.

File: cpp/src/plasma/common/common_extension.cc [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+PyObject *CommonError;
+
+/* Initialize pickle module. */
+
+PyObject *pickle_module = NULL;
+PyObject *pickle_loads = NULL;
+PyObject *pickle_dumps = NULL;
+PyObject *pickle_protocol = NULL;
+
+void init_pickle_module(void) {
```
Comment (robertnishihara) [inline cpp/src/plasma/common/common_extension.cc]: I think this module makes more sense as part of Ray. Maybe the only part that is needed in Plasma is `PyStringToUniqueID`, which would be fine to duplicate.

File: cpp/src/plasma/common/io.cc [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+#include "common.h"
+#include "event_loop.h"
+#include "arrow/util/logging.h"
+
+#ifndef _WIN32
+/* This function is actually not declared in standard POSIX, so declare it. */
+extern int usleep(useconds_t usec);
+#endif
+
```
Comment (robertnishihara) [inline cpp/src/plasma/common/io.cc]: This file is definitely needed in Ray as well, so what's the plan for that? Should we just duplicate it?

File: cpp/src/plasma/format/plasma_generated.h [hunk 4]
```diff
@@ -0,0 +1,2999 @@
+// automatically generated by the FlatBuffers compiler, do not modify
```
Comment (robertnishihara) [inline cpp/src/plasma/format/plasma_generated.h]: You probably didn't mean to commit this, right?

File: cpp/src/plasma/malloc.c [hunk 5]
```diff
@@ -0,0 +1,170 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
```
Comment (robertnishihara) [inline cpp/src/plasma/malloc.c]: Are you potentially modifying this file in https://github.com/ray-project/ray/pull/411?

File: cpp/src/plasma/plasma_extension.cc [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+  int has_object;
+  plasma_contains(conn, object_id, &has_object);
+
+  if (has_object)
+    Py_RETURN_TRUE;
+  else
+    Py_RETURN_FALSE;
+}
+
+PyObject *PyPlasma_fetch(PyObject *self, PyObject *args) {
```
Comment (robertnishihara) [inline cpp/src/plasma/plasma_extension.cc]: I assume we're moving the plasma store and not the plasma manager to Arrow right now, so some of these won't actually compile and should be removed (`PyPlasma_fetch`, `PyPlasma_wait`). Also `PyPlasma_connect` will have to be changed to work without a manager.

File: cpp/src/plasma/test/run_tests.sh [hunk 7]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+
+# Cause the script to exit if a single command fails.
+set -e
```
Comment (robertnishihara) [inline cpp/src/plasma/test/run_tests.sh]: These tests will need to change (but should still be run).

**Review by robertnishihara — COMMENTED** [review]
This is a good start. The vast majority of the testing of plasma is done in `ray/python/ray/plasma/test/test.py`, so we should bring that over, strip out the manager stuff, and run it in the CI.

**Comment by robertnishihara** [comment]
I had a look through the PR, and I think the majority of the work will be the following two things:

1. Separating out the plasma store code from the plasma manager code. There shouldn't be any conceptual problems here, but it will take some work.
2. Building/testing Plasma in the Arrow CI. We obviously don't want to slow down the Arrow CI, but we would like to test all four combinations of OS X and Linux with Python 2 and Python 3, and separately test things in valgrind as well as linting. Currently in Ray we're using 6 Travis builds, each of which takes about 20 minutes. So we'll need to figure out a good approach here.

File: cpp/src/plasma/cmake/Common.cmake [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+message(STATUS "Flatbuffers include dir: ${FLATBUFFERS_INCLUDE_DIR}")
+message(STATUS "Flatbuffers static library: ${FLATBUFFERS_STATIC_LIB}")
+message(STATUS "Flatbuffers compiler: ${FLATBUFFERS_COMPILER}")
+include_directories(SYSTEM ${FLATBUFFERS_INCLUDE_DIR})
+
+# Custom CFLAGS
+
+set(CMAKE_C_FLAGS "-g -Wall -Wextra -Werror=implicit-function-declaration -Wno-sign-compare -Wno-unused-parameter -Wno-type-limits -Wno-missing-field-initializers --std=c99 -D_XOPEN_SOURCE=500 -D_POSIX_C_SOURCE=200809L -fPIC -std=c99")
+
+# Code for finding Python
```
Comment (wesm) [inline cpp/src/plasma/cmake/Common.cmake]: There's https://github.com/apache/arrow/blob/master/cpp/cmake_modules/FindPythonLibsNew.cmake, is one or the other made redundant by the other?

**Comment by wesm** [comment]
Let's wait to merge into the branch until you can incorporate the in-flight C++ changes happening? 

Do you have a sense of what's taking a lot of time in your Travis CI builds (the tests themselves, or something else)? We're already running OS X and Linux for both 2.7 and 3.5 (with 3.6 being added soon). We could set up separate entries in the build matrix for the object store if the runtime ends up being too long

**Comment by robertnishihara** [comment]
Yeah, we'll include the in-flight C++ changes in this PR.

In our tests, we spend a couple minutes downloading dependencies (e.g., Boost) and building Ray. However, most of these are integration tests, and the tests specific to the object store only take one or two minutes.

**Comment by wesm** [comment]
Ok, that probably won't impact the Arrow build times too much then -- if you build against Arrow master in Ray (and set a particular release or commit for releases) then that is probably sufficient for ongoing integration testing.

**Comment by wesm** [comment]
After 0.3, if you can update this with Ray's latest we can get it merged into a branch so we can sort out C++ porting / testing / CI, etc?

**Comment by robertnishihara** [comment]
Yes, that sounds good!

**Comment by pcmoritz** [comment]
replaced by https://github.com/apache/arrow/issues/742

**Pull request closed without merging**